File: src/libclr-common/DatabaseNode.cs
namespace Petroules.Silverlock
{
    using System;
    using System.Xml;

    public abstract class DatabaseNode
    {
        private Group parent;
        private DateTime created;
        private DateTime accessed;
        private DateTime modified;
        private Guid uuid;
        private string title;

        protected DatabaseNode(string title = "")
        {
            this.created = DateTime.UtcNow;
            this.accessed = DateTime.UtcNow;
            this.modified = DateTime.UtcNow;
            this.uuid = Guid.NewGuid();
            this.title = title;
        }

        public abstract void Destroy();

        public DateTime Created
        {
            get
            {
                return this.created.ToUniversalTime();
            }

            private set
            {
                if (this.created != value)
                {
                    this.created = value.ToUniversalTime();
                }
            }
        }

        public DateTime Accessed
        {
            get
            {
                return this.accessed.ToUniversalTime();
            }

            private set
            {
                if (this.accessed != value)
                {
                    this.accessed = value.ToUniversalTime();
                }
            }
        }

        public DateTime Modified
        {
            get
            {
                return this.modified.ToUniversalTime();
            }

            protected set
            {
                if (this.modified != value)
                {
                    this.modified = value.ToUniversalTime();
                }
            }
        }

        public Guid UUID
        {
            get
            {
                return this.uuid;
            }

            protected set
            {
                if (this.uuid != value)
                {
                    this.uuid = value;
                    this.SetModified();
                }
            }
        }

        public string Title
        {
            get
            {
                return this.title;
            }

            set
            {
                if (this.title != value)
                {
                    this.title = value;
                    this.SetModified();
                }
            }
        }

        public virtual Group ParentNode
        {
            get
            {
                return this.parent;
            }

            set
            {
                // Cant't self-parent or parent to the same node
                if (this == value || this.parent == value)
                {
                    return;
                }

                this.Detach();
                this.parent = value;
                this.Attach();
            }
        }

        public Group RootNode
        {
            get
            {
                if (!this.IsAttached)
                {
                    return null;
                }

                Group parent = this.parent;
                while (parent.ParentNode != null)
                {
                    parent = parent.ParentNode;
                }

                return parent;
            }
        }

        public bool HasDatabase
        {
            get { return this.RootNode as Database != null; }
        }

        public bool IsAttached
        {
            get { return this.ParentNode != null; }
        }

        public event EventHandler TreeModified = delegate { };

        public void SetAccessed()
        {
            this.Accessed = DateTime.UtcNow;
        }

        public abstract XmlElement ToXml(XmlDocument document);

        protected XmlElement ToXml(XmlDocument document, string tagName)
        {
            XmlElement element = document.CreateElement(tagName);
            element.SetAttribute(DatabaseKeys.XML_CREATED, this.Created.ToUniversalTime().ToString("") + "Z");
            element.SetAttribute(DatabaseKeys.XML_ACCESSED, this.Accessed.ToUniversalTime().ToString("") + "Z");
            element.SetAttribute(DatabaseKeys.XML_MODIFIED, this.Modified.ToUniversalTime().ToString("") + "Z");
            element.SetAttribute(DatabaseKeys.XML_UUID, this.UUID.ToString());
            element.SetAttribute(DatabaseKeys.XML_TITLE, this.Title);
            return element;
        }

        protected void SetModified()
        {
            this.Modified = DateTime.UtcNow;
        }

        protected void Detach()
        {
            if (this.IsAttached)
            {
                this.DetachFromList();
                DatabaseNode parentTemp = this.parent;
                this.parent = null;

                // Notify the tree that it has been modified and
                // disconnect our modification event from the parent's
                this.TreeModified(this, EventArgs.Empty);
                this.TreeModified -= parentTemp.TreeModified;
            }
        }

        protected internal virtual void FromXml(XmlElement element)
        {
            if (this.parent != null)
            {
                throw new InvalidOperationException("Cannot read a DatabaseNode from an XML element if it has a parent");
            }

            this.Created = DateTime.Parse(element.Attributes[DatabaseKeys.XML_CREATED].Value);
            this.Accessed = DateTime.Parse(element.Attributes[DatabaseKeys.XML_ACCESSED].Value);
            this.Modified = DateTime.Parse(element.Attributes[DatabaseKeys.XML_MODIFIED].Value);
            this.UUID = Guid.Parse(element.Attributes[DatabaseKeys.XML_UUID].Value);
            this.Title = element.Attributes[DatabaseKeys.XML_TITLE].Value;
        }

        protected abstract void AttachToList();

        protected abstract void DetachFromList();

        private void Attach()
        {
            if (this.parent != null)
            {
                // Ensure UUID uniquity
                while (this.RootNode.ContainsUUID(this.UUID))
                {
                    this.UUID = Guid.NewGuid();
                }

                // When we add an entry to the parent, the DB was modified
                this.AttachToList();

                // Connect our modification event to the parent's
                // and notify the tree that it has been modified
                this.TreeModified += this.parent.TreeModified;
                this.TreeModified(this, EventArgs.Empty);
            }
        }
    }
}

File: src/lib/java/src/com/petroules/silverlock/DatabaseNode.java
package com.petroules.silverlock;

import java.text.DateFormat;
import java.text.ParseException;
import java.util.Date;
import java.util.Observable;
import java.util.Observer;
import java.util.UUID;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

public abstract class DatabaseNode extends Observable implements Observer
{
	private Group parent;
	private Date created;
	private Date accessed;
	private Date modified;
	private UUID uuid;
	private String title;

	protected DatabaseNode()
	{
		this("");
	}

	protected DatabaseNode(String title)
	{
		this.created = new Date();
		this.accessed = new Date();
		this.modified = new Date();
		this.uuid = UUID.randomUUID();
		this.title = title;
	}

	public abstract void destroy();

	public Date getCreated()
	{
		return this.created;
	}

	private void setCreated(Date value)
	{
		if (!this.created.equals(value))
		{
			this.created = value;
		}
	}

	public Date getAccessed()
	{
		return this.accessed;
	}

	private void setAccessed(Date value)
	{
		if (!this.accessed.equals(value))
		{
			this.accessed = value;
		}
	}

	public Date getModified()
	{
		return this.modified;
	}

	protected void setModified(Date value)
	{
		if (!this.modified.equals(value))
		{
			this.modified = value;
		}
	}

	public UUID getUUID()
	{
		return this.uuid;
	}

	protected void setUUID(UUID value)
	{
		if (!this.uuid.equals(value))
		{
			this.uuid = value;
			this.setModified();
		}
	}

	public String getTitle()
	{
		return this.title;
	}

	public void setTitle(String value)
	{
		if (!this.title.equals(value))
		{
			this.title = value;
			this.setModified();
		}
	}

	public Group getParentNode()
	{
		return this.parent;
	}

	public void setParentNode(Group value)
	{
		// Cant't self-parent or parent to the same node
		if (this == value || this.parent == value)
		{
			return;
		}

		this.detach();
		this.parent = value;
		this.attach();
	}

	public Group getRootNode()
	{
		if (!this.isAttached())
		{
			return null;
		}

		Group parent = this.parent;
		while (parent.getParentNode() != null)
		{
			parent = parent.getParentNode();
		}

		return parent;
	}

	public boolean hasDatabase()
	{
		return this.getRootNode() instanceof Database;
	}

	public boolean isAttached()
	{
		return this.getParentNode() != null;
	}

	public void setAccessed()
	{
		this.setAccessed(new Date());
	}

	public abstract Element toXml(Document document);

	protected Element toXml(Document document, String tagName)
	{
		Element element = document.createElement(tagName);
		element.setAttribute(DatabaseKeys.XML_CREATED, this.getCreated().toString() + "Z");
		element.setAttribute(DatabaseKeys.XML_ACCESSED, this.getAccessed().toString() + "Z");
		element.setAttribute(DatabaseKeys.XML_MODIFIED, this.getModified().toString() + "Z");
		element.setAttribute(DatabaseKeys.XML_UUID, uuidToString(this.getUUID()));
		element.setAttribute(DatabaseKeys.XML_TITLE, this.getTitle());
		return element;
	}

	protected void setModified()
	{
		this.setModified(new Date());
	}

	protected void detach()
	{
		if (this.isAttached())
		{
			this.detachFromList();
			DatabaseNode parentTemp = this.parent;
			this.parent = null;

			// Notify the tree that it has been modified and
			// disconnect our modification event from the parent's
			this.setChanged();
			this.notifyObservers();
			this.deleteObserver(parentTemp);
		}
	}

	protected/* internal */void fromXml(Element element)
	{
		if (this.parent != null)
		{
			throw new IllegalStateException("Cannot read a DatabaseNode from an XML element if it has a parent");
		}

		DateFormat dateFormat = DateFormat.getDateTimeInstance();
		try
		{
			this.setCreated(dateFormat.parse(element.getAttribute(DatabaseKeys.XML_CREATED)));
		}
		catch (ParseException e)
		{
		}

		try
		{
			this.setAccessed(dateFormat.parse(element.getAttribute(DatabaseKeys.XML_ACCESSED)));
		}
		catch (ParseException e)
		{
		}

		try
		{
			this.setModified(dateFormat.parse(element.getAttribute(DatabaseKeys.XML_MODIFIED)));
		}
		catch (ParseException e)
		{
		}

		this.setUUID(uuidFromString(element.getAttribute(DatabaseKeys.XML_UUID)));
		this.setTitle(element.getAttribute(DatabaseKeys.XML_TITLE));
	}

	protected abstract void attachToList();

	protected abstract void detachFromList();

	private void attach()
	{
		if (this.parent != null)
		{
			// Ensure UUID uniquity
			while (this.getRootNode().containsUUID(this.getUUID()))
			{
				this.setUUID(UUID.randomUUID());
			}

			// When we add an entry to the parent, the DB was modified
			this.attachToList();

			// Connect our modification event to the parent's
			// and notify the tree that it has been modified
			this.addObserver(this.parent);
			this.setChanged();
			this.notifyObservers();
		}
	}

	@Override
	public void update(Observable o, Object arg)
	{
	}
	
	private String uuidToString(UUID u)
	{
		return "{" + u + "}";
	}
	
	private UUID uuidFromString(String s)
	{
		return UUID.fromString(s.replace("{", "").replace("}", ""));
	}
}
